lib/chef/resource/locale.rb

```diff
@@ -25,65 +25,91 @@ class Locale < Chef::Resource
+        end
+      end
 
       action :update do
         description "Update the system's locale."
-
-        if node["init_package"] == "systemd"
-          # on systemd settings LC_ALL is (correctly) reserved only for testing and cannot be set globally
-          execute "localectl set-locale LANG=#{new_resource.lang}" do
-            # RHEL uses /etc/locale.conf
-            not_if { up_to_date?("/etc/locale.conf", new_resource.lang) } if ::File.exist?("/etc/locale.conf")
-            # Ubuntu 16.04 still uses /etc/default/locale
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang) } if ::File.exist?("/etc/default/locale")
+        begin
+          unless up_to_date?
+            converge_by "Updating System Locale" do
+              generate_locales unless unavailable_locales.empty?
+              update_locale
+            end
           end
-        elsif ::File.exist?("/etc/sysconfig/i18n")
-          locale_file_path = "/etc/sysconfig/i18n"
-
-          updated = up_to_date?(locale_file_path, new_resource.lang, new_resource.lc_all)
+        rescue
+          raise "#{node['platform']} platform not supported by the chef locale resource."
+        end
+      end
 
-          file locale_file_path do
-            content lazy {
-              locale = IO.read(locale_file_path)
-              variables = Hash[locale.lines.map { |line| line.strip.split("=") }]
-              variables["LANG"] = new_resource.lang
-              variables["LC_ALL"] =
-                variables.map { |pairs| pairs.join("=") }.join("\n") + "\n"
-            }
-            not_if { updated }
+      action_class do
+        def generate_locales
+          bash "Generating locales: #{unavailable_locales.join(' ')}" do
+            code "locale-gen #{unavailable_locales.join(' ')}"
           end
+        end
 
-          execute "reload root's lang profile script" do
-            command "source /etc/sysconfig/i18n; source /etc/profile.d/lang.sh"
-            not_if { updated }
+        def update_locale
+          file "Updating system locale" do
+            path LOCALE_CONF
+            content new_content
           end
-        elsif ::File.exist?("/usr/sbin/update-locale")
-          execute "Generate locales" do
-            command "locale-gen"
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang, new_resource.lc_all) }
+          bash "Reload root's lang profile script" do
+            code "source #{LOCALE_CONF} && export $(cut -d= -f1 #{LOCALE_CONF})"
+            only_if { ::File.exist?(LOCALE_CONF) }
           end
+        end
 
-          execute "Update locale" do
-            command "update-locale LANG=#{new_resource.lang} LC_ALL=#{new_resource.lc_all}"
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang, new_resource.lc_all) }
+        def unavailable_locales
```

Comment:
It will be good if we document what this method does? Easy to understand to people who read it ?